app/src/main/java/org/oppia/android/app/story/storyitemviewmodel/StoryChapterSummaryViewModel.kt

```diff
@@ -33,16 +38,100 @@ class StoryChapterSummaryViewModel(
   val summary: String = chapterSummary.summary
   val chapterThumbnail: LessonThumbnail = chapterSummary.chapterThumbnail
   val missingPrerequisiteChapter: ChapterSummary = chapterSummary.missingPrerequisiteChapter
+  val chapterPlayState: ChapterPlayState = chapterSummary.chapterPlayState
 
   fun onExplorationClicked() {
+    val shouldSavePartialProgress =
+      when (chapterPlayState) {
+        ChapterPlayState.IN_PROGRESS_SAVED, ChapterPlayState.IN_PROGRESS_NOT_SAVED,
+        ChapterPlayState.STARTED_NOT_COMPLETED, ChapterPlayState.NOT_STARTED -> true
+        else -> false
+      }
+    if (chapterPlayState == ChapterPlayState.IN_PROGRESS_SAVED) {
+      explorationCheckpointController.isSavedCheckpointCompatibleWithExploration(
+        ProfileId.getDefaultInstance(),
+        explorationId
+      ).toLiveData().observe(
+        fragment,
+        Observer {
+          if (it.isSuccess()) {
+            startOrResumeExploration(
+              internalProfileId,
+              topicId,
+              storyId,
+              explorationId,
+              shouldSavePartialProgress,
+              canExplorationBeResumed = it.getOrThrow(),
+              backflowScreen = 1
+            )
+          } else if (it.isFailure()) {
+            startOrResumeExploration(
+              internalProfileId,
+              topicId,
+              storyId,
+              explorationId,
+              shouldSavePartialProgress,
+              canExplorationBeResumed = false,
+              backflowScreen = 1
+            )
+          }
+        }
+      )
+    } else {
+      startOrResumeExploration(
+        internalProfileId,
+        topicId,
+        storyId,
+        explorationId,
+        shouldSavePartialProgress,
+        canExplorationBeResumed = false,
+        backflowScreen = 1
+      )
+    }
+  }
+
+  private fun startOrResumeExploration(
+    internalProfileId: Int,
+    topicId: String,
+    storyId: String,
+    explorationId: String,
+    shouldSavePartialProgress: Boolean,
+    canExplorationBeResumed: Boolean,
+    backflowScreen: Int?
+  ) {
+    if (canExplorationBeResumed) {
+      explorationSelectionListener.selectExploration(
+        internalProfileId,
+        topicId,
+        storyId,
+        explorationId,
+        canExplorationBeResumed,
+        shouldSavePartialProgress,
+        backflowScreen
+      )
+    } else {
+      playExploration(
+        canExplorationBeResumed,
+        shouldSavePartialProgress,
+        backflowScreen
+      )
+    }
+  }
```

Comment:
If we need to check at all locations whether there is compatibility, and in that case we just abandon the checkpoint & start over, why not move the version check into the DataProvider that provides the checkpoint, return a default, and short-circuit to ExplorationActivity if no checkpoint was found?

Reply:
As per our chat, used `removeObserver`